File: lib/models/travel/booking_types.dart
enum BookingType {
  flight,
  hotel,
  train,
  bus,
  event,
  attraction,
  visa,
  vacationRental;

  static BookingType fromString(String value) {
    switch (value) {
      case 'flight':
        return BookingType.flight;
      case 'hotel':
        return BookingType.hotel;
      case 'train':
        return BookingType.train;
      case 'bus':
        return BookingType.bus;
      case 'event':
        return BookingType.event;
      case 'attraction':
        return BookingType.attraction;
      case 'visa':
        return BookingType.visa;
      case 'vacation_rental':
        return BookingType.vacationRental;
      default:
        return BookingType.flight;
    }
  }

  String toJson() {
    switch (this) {
      case BookingType.vacationRental:
        return 'vacation_rental';
      default:
        return name;
    }
  }
}

class Amount {
  final double? amount;
  final String? currency;

  Amount({this.amount, this.currency});

  factory Amount.fromJson(Map<String, dynamic>? json) {
    if (json == null) return Amount();
    return Amount(
      amount: (json['amount'] as num?)?.toDouble(),
      currency: json['currency'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'amount': amount,
        'currency': currency,
      };
}

class ContactInformation {
  final String? email;
  final String? phone;

  ContactInformation({this.email, this.phone});

  factory ContactInformation.fromJson(Map<String, dynamic>? json) {
    if (json == null) return ContactInformation();
    return ContactInformation(
      email: json['email'] as String?,
      phone: json['phone'] as String?,
    );
  }
}

class Address {
  final String? street;
  final String? city;
  final String? state;
  final String? country;
  final String? postalCode;

  Address({this.street, this.city, this.state, this.country, this.postalCode});

  factory Address.fromJson(Map<String, dynamic>? json) {
    if (json == null) return Address();
    return Address(
      street: json['street'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      country: json['country'] as String?,
      postalCode: json['postal_code'] as String?,
    );
  }
}

class Passenger {
  final String? name;
  final String? type;
  final String? ticketNumber;
  final String? seatNumber;
  final int? age;
  final String? gender;

  Passenger({
    this.name,
    this.type,
    this.ticketNumber,
    this.seatNumber,
    this.age,
    this.gender,
  });

  factory Passenger.fromJson(Map<String, dynamic> json) {
    return Passenger(
      name: json['name'] as String?,
      type: json['type'] as String?,
      ticketNumber: json['ticket_number'] as String?,
      seatNumber: json['seat_number'] as String?,
      age: json['age'] as int?,
      gender: json['gender'] as String?,
    );
  }
}

class FlightLocation {
  final String? city;
  final String? airport;
  final String? airportCode;
  final String? terminal;
  final String? datetime;
  final double? latitude;
  final double? longitude;

  FlightLocation({
    this.city,
    this.airport,
    this.airportCode,
    this.terminal,
    this.datetime,
    this.latitude,
    this.longitude,
  });

  factory FlightLocation.fromJson(Map<String, dynamic>? json) {
    if (json == null) return FlightLocation();
    return FlightLocation(
      city: json['city'] as String?,
      airport: json['airport'] as String?,
      airportCode: json['airport_code'] as String?,
      terminal: json['terminal'] as String?,
      datetime: json['datetime'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
    );
  }
}

class FlightSegment {
  final String? flightNumber;
  final String? airline;
  final String? operatedBy;
  final FlightLocation departure;
  final FlightLocation arrival;
  final String? flightClass;
  final String? fareType;
  final String? duration;
  final Map<String, String?>? baggageAllowance;

  FlightSegment({
    this.flightNumber,
    this.airline,
    this.operatedBy,
    required this.departure,
    required this.arrival,
    this.flightClass,
    this.fareType,
    this.duration,
    this.baggageAllowance,
  });

  factory FlightSegment.fromJson(Map<String, dynamic> json) {
    return FlightSegment(
      flightNumber: json['flight_number'] as String?,
      airline: json['airline'] as String?,
      operatedBy: json['operated_by'] as String?,
      departure: FlightLocation.fromJson(json['departure'] as Map<String, dynamic>?),
      arrival: FlightLocation.fromJson(json['arrival'] as Map<String, dynamic>?),
      flightClass: json['class'] as String?,
      fareType: json['fare_type'] as String?,
      duration: json['duration'] as String?,
      baggageAllowance: json['baggage_allowance'] != null
          ? Map<String, String?>.from(json['baggage_allowance'])
          : null,
    );
  }
}

class FareBreakdown {
  final Amount? baseFare;
  final Amount? taxes;
  final Amount? fees;
  final Amount? discounts;
  final Amount total;

  FareBreakdown({
    this.baseFare,
    this.taxes,
    this.fees,
    this.discounts,
    required this.total,
  });

  factory FareBreakdown.fromJson(Map<String, dynamic>? json) {
    if (json == null) return FareBreakdown(total: Amount());
    return FareBreakdown(
      baseFare: Amount.fromJson(json['base_fare'] as Map<String, dynamic>?),
      taxes: Amount.fromJson(json['taxes'] as Map<String, dynamic>?),
      fees: Amount.fromJson(json['fees'] as Map<String, dynamic>?),
      discounts: Amount.fromJson(json['discounts'] as Map<String, dynamic>?),
      total: Amount.fromJson(json['total'] as Map<String, dynamic>?),
    );
  }
}

File: functions/src/types/bookingTypes.ts
export type BookingType = 'flight' | 'hotel' | 'train' | 'bus' | 'event' | 'attraction' | 'visa' | 'vacation_rental';

export interface Amount {
  amount: number | null;
  currency: string | null;
}

export interface Passenger {
  name: string | null;
  type: string | null;
  ticket_number?: string | null;
  seat_number?: string | null;
  meal_preference?: string | null;
  age?: number | null;
  gender?: string | null;
}

export interface FlightLocation {
  city: string | null;
  airport: string | null;
  airport_code: string | null;
  terminal: string | null;
  datetime: string | null;
  latitude?: number | null;
  longitude?: number | null;
}

export interface FlightSegment {
  flight_number: string | null;
  airline: string | null;
  operated_by?: string | null;
  departure: FlightLocation;
  arrival: FlightLocation;
  class: string | null;
  fare_type: string | null;
  duration: string | null;
  baggage_allowance?: {
    checkin: string | null;
    cabin: string | null;
  };
  seat_assignments?: Array<{
    passenger_name: string | null;
    seat_number: string | null;
  }>;
}

export interface FareBreakdown {
  base_fare?: Amount;
  room_rate?: Amount;
  nightly_rate?: Amount;
  cleaning_fee?: Amount;
  service_fee?: Amount;
  taxes?: Amount;
  fees?: Amount;
  discounts?: Amount;
  reservation_charges?: Amount;
  booking_fees?: Amount;
  ticket_price?: Amount;
  total: Amount;
}

export interface ContactInformation {
  email: string | null;
  phone: string | null;
}

export interface Address {
  street?: string | null;
  city: string | null;
  state?: string | null;
  country?: string | null;
  postal_code?: string | null;
}

export interface BaseBooking {
  booking_type: BookingType;
  booking_reference?: string | null;
  booking_date?: string | null;
  status?: string | null;
  contact_information?: ContactInformation;
  date?: string;
  pdfParseStatus?: string;
  pdfParseError?: string;
  origin?: string;
  destination?: string;
  amount?: number;
  currency?: string;
  amountINR?: number;
}

export interface FlightBooking extends BaseBooking {
  booking_type: 'flight';
  airline?: string;
  trip_type?: string | null;
  total_amount_paid?: Amount;
  passengers?: Passenger[];
  flights?: FlightSegment[];
  fare_breakdown?: FareBreakdown;
  additional_info?: string | null;
}

export interface HotelBooking extends BaseBooking {
  booking_type: 'hotel';
  hotel_name: string;
  hotel_brand?: string | null;
  address: Address;
  check_in: string;
  check_out: string;
  nights: number;
  room_details?: {
    room_type?: string | null;
    bed_type?: string | null;
    number_of_rooms?: number | null;
    number_of_guests?: number | null;
  };
  guests?: Passenger[];
  total_amount: Amount;
  fare_breakdown?: FareBreakdown;
  amenities?: string[];
  cancellation_policy?: string | null;
}

export interface TrainBooking extends BaseBooking {
  booking_type: 'train';
  pnr: string;
  train_number: string;
  train_name: string;
  departure: {
    station: string | null;
    station_code: string | null;
    city: string | null;
    datetime: string | null;
    platform?: string | null;
  };
  arrival: {
    station: string | null;
    station_code: string | null;
    city: string | null;
    datetime: string | null;
    platform?: string | null;
  };
  duration: string | null;
  class: string | null;
  coach?: string | null;
  seat_berth?: string | null;
  passengers?: Passenger[];
  total_amount: Amount;
  fare_breakdown?: FareBreakdown;
}

export interface BusBooking extends BaseBooking {
  booking_type: 'bus';
  bus_operator: string;
  bus_type?: string | null;
  departure: {
    location: string | null;
    city: string | null;
    datetime: string | null;
  };
  arrival: {
    location: string | null;
    city: string | null;
    datetime: string | null;
  };
  duration: string | null;
  seat_numbers?: string[];
  passengers?: Passenger[];
  total_amount: Amount;
  amenities?: string[];
  boarding_point?: string | null;
  dropping_point?: string | null;
}

export interface EventBooking extends BaseBooking {
  booking_type: 'event';
  event_name: string;
  event_type?: string | null;
  venue: {
    name: string | null;
    address: string | null;
    city: string | null;
    country: string | null;
  };
  event_date: string;
  doors_open?: string | null;
  seats?: Array<{
    section?: string | null;
    row?: string | null;
    seat_number?: string | null;
    ticket_holder?: string | null;
  }>;
  number_of_tickets: number;
  total_amount: Amount;
  fare_breakdown?: FareBreakdown;
  barcode_qr?: string | null;
}

export interface AttractionBooking extends BaseBooking {
  booking_type: 'attraction';
  attraction_name: string;
  attraction_type?: string | null;
  location: {
    name?: string | null;
    address?: string | null;
    city: string | null;
    country?: string | null;
  };
  visit_date: string;
  visit_time?: string | null;
  duration?: string | null;
  tickets?: Array<{
    type: string | null;
    quantity: number | null;
    holder_name?: string | null;
  }>;
  total_amount: Amount;
  includes?: string[];
  meeting_point?: string | null;
}

export interface VisaBooking extends BaseBooking {
  booking_type: 'visa';
  application_reference: string;
  application_date: string;
  visa_type: string;
  country: string;
  applicant: {
    name: string | null;
    passport_number: string | null;
    nationality: string | null;
    date_of_birth: string | null;
  };
  visa_details?: {
    visa_number?: string | null;
    validity_from?: string | null;
    validity_to?: string | null;
    duration?: string | null;
    entries?: string | null;
  };
  appointment?: {
    date: string | null;
    time: string | null;
    location: string | null;
    address: string | null;
  };
  fees: Amount;
}

export interface VacationRentalBooking extends BaseBooking {
  booking_type: 'vacation_rental';
  property_name: string;
  property_type?: string | null;
  host_name?: string | null;
  address: Address;
  check_in: string;
  check_out: string;
  nights: number;
  guests: number;
  total_amount: Amount;
  fare_breakdown?: FareBreakdown;
  amenities?: string[];
  house_rules?: string | null;
  cancellation_policy?: string | null;
}

export type Booking =
  | FlightBooking
  | HotelBooking
  | TrainBooking
  | BusBooking
  | EventBooking
  | AttractionBooking
  | VisaBooking
  | VacationRentalBooking;

export interface TravelAnalysisResponse {
  travels: Booking[];
  moreBatches: boolean;
  nextBatch: number | null;
  totalEmails?: number;
}

export interface AnalyzeTravelRequest {
  accessToken: string;
  options?: {
    batchSize?: number;
    batch?: number;
    year?: number;
  };
}
